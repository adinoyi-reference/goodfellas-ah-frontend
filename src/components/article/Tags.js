import React, { Component } from 'react';
import { connect } from 'react-redux';
import swal from 'sweetalert';
import Button from '../shared/Button';
import { addTags } from '../../actions/articleActions';
import publishArticle from '../../actions/publishArticle';

class Tags extends Component {
  state = {
    tagsCount: 0,
    tags: [],
    articleStatus: false,
  };

  handleSubmit = (e) => {
    // First prevent default form submission behaviour
    e.preventDefault();

    //  Check for empty spaces and process accordingly
    const newTag = this.refs.getTags.value;
    const { tagsCount, tags } = this.state;

    if (newTag.trim() !== '') {
      this.setState({
        tags: [...tags, { tag: newTag.toLowerCase().trim(), tagsCount: tagsCount + 1 }],
        tagsCount: tagsCount + 1,
      });
    }

    // Reset input field
    e.target.reset();
  };

  handleClick = (e) => {
    const { tags } = this.state;
    const currentTag = e.target.previousSibling.textContent;
    const newState = [];

    // Check for the tag being deleted in the state

    for (let index = 0; index < tags.length; index += 1) {
      const element = tags[index];
      // create a new state without that tag in it

      if (element.tag !== currentTag.trim()) {
        newState.push({ tag: element.tag, tagsCount: element.tagsCount });
      }
    }

    // update the state with the new state object
    this.setState({
      tags: [...newState],
    });
  };

  componentDidUpdate = () => {
    const { tags, articleStatus } = this.state;
    const { publishedArticle, history } = this.props;

    // Check if the limit for tags has been reached
    if (tags.length === 5) {
      this.refs.getTags.disabled = true;
      this.refs.getTags.placeholder = 'You\'ve reached your limit';
    } else {
      this.refs.getTags.disabled = false;
      this.refs.getTags.placeholder = 'Add a tag...';
    }

    // Check if article has been published then publish Tags
    if (publishedArticle) {
      this.publishTags();
    }

    if (articleStatus) {
      swal('Good job!', 'Article Saved Successfully!', 'success');
      setTimeout(() => history.push(`/articles/${publishedArticle.slug}`), 3000);
    }
  };

  publishArticle = async () => {
    const { location, publishArticle: publishNewArticle } = this.props;

    await publishNewArticle(location.state.articlePayload);

    this.setState({ articleStatus: true });
  };

  publishTags = async () => {
    const { tags } = this.state;
    const {
      addTags: addNewTags,
      publishedArticle,
    } = this.props;
    const tagsList = tags.map((eachTag) => {
      const { tag } = eachTag;
      return tag;
    });

    await addNewTags(tagsList, publishedArticle.slug);
  };

  render() {
    const { tags } = this.state;
    return (
      <div className="tags">
        <div className="tags-main">
          <h2>Add tags to your article</h2>
          <div className="tags-main-details">
            <p>
              Help your target audience find this article by adding up to 5 tags
            </p>
            <form onSubmit={this.handleSubmit}>
              <input ref="getTags" type="text" placeholder="Add a tag..." />
            </form>
            <div className="tags-wrapper">
              {tags.map((eachTag) => {
                const { tagsCount, tag } = eachTag;
                return (
                  <div key={tagsCount} className="new-tag">
                    <p>{tag} </p>
                    <span onClick={this.handleClick} className="new-tag-cancel">
                      x
                    </span>
                  </div>
                );
              })}
            </div>
            <div onClick={this.publishArticle}>
              <Button
                className="btn hero-section-greenbutton"
                type="submit"
                title="Publish"
              />
            </div>
          </div>
        </div>
      </div>
    );
  }
}


function mapStatetoProps(state) {
  return {
    tags: state.articles.tags,
    tagsError: state.articles.tagsError,
    publishedArticle: state.publishArticleReducer.publishedArticle.article,
  };
}

export default connect(mapStatetoProps, { addTags, publishArticle })(Tags);
