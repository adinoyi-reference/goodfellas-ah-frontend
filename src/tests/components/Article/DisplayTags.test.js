import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import toJson from 'enzyme-to-json';
import Tags from '../../../components/article/DisplayTags';
import Root from '../../../root';


const articles = {
  tagList: ['tag', 'tags'],
};

const mockDisplayTags = mount(
  <Root>
    <MemoryRouter initialEntries={[{ key: 'testkey' }]}>
      <Tags articles={articles} />
    </MemoryRouter>
  </Root>,
);

describe('Display Tags UI', () => {
  describe('render features', () => {
    test('component should render as expected', () => {
      const tree = toJson(mockDisplayTags);
      expect(tree).toMatchSnapshot();
    });
  });
});
