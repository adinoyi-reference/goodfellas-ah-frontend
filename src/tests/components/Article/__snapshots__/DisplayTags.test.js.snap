// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display Tags UI render features component should render as expected 1`] = `
<_default>
  <Provider
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  >
    <MemoryRouter
      initialEntries={
        Array [
          Object {
            "key": "testkey",
          },
        ]
      }
    >
      <Router
        history={
          Object {
            "action": "POP",
            "block": [Function],
            "canGo": [Function],
            "createHref": [Function],
            "entries": Array [
              Object {
                "hash": "",
                "key": "testkey",
                "pathname": "/",
                "search": "",
              },
            ],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "index": 0,
            "length": 1,
            "listen": [Function],
            "location": Object {
              "hash": "",
              "key": "testkey",
              "pathname": "/",
              "search": "",
            },
            "push": [Function],
            "replace": [Function],
          }
        }
      >
        <DisplayTags
          articles={
            Object {
              "tagList": Array [
                "tag",
                "tags",
              ],
            }
          }
        >
          <div
            className="tags-display"
          >
            <p
              className="new-tag"
              key="1"
            >
              tag
            </p>
            <p
              className="new-tag"
              key="2"
            >
              tags
            </p>
          </div>
        </DisplayTags>
      </Router>
    </MemoryRouter>
  </Provider>
</_default>
`;
